README.md

ASN.1 library for Python
------------------------
[![PyPI](https://img.shields.io/pypi/v/pyasn1.svg?maxAge=2592000)](https://pypi.org/project/pyasn1)
[![Python Versions](https://img.shields.io/pypi/pyversions/pyasn1.svg)](https://pypi.org/project/pyasn1/)
[![Build status](https://github.com/pyasn1/pyasn1/actions/workflows/main.yml/badge.svg)](https://github.com/pyasn1/pyasn1/actions/workflows/main.yml)
[![Coverage Status](https://img.shields.io/codecov/c/github/pyasn1/pyasn1.svg)](https://codecov.io/github/pyasn1/pyasn1)
[![GitHub license](https://img.shields.io/badge/license-BSD-blue.svg)](https://raw.githubusercontent.com/pyasn1/pyasn1/master/LICENSE.txt)

This is a free and open source implementation of ASN.1 types and codecs
as a Python package. It has been first written to support particular
protocol (SNMP) but then generalized to be suitable for a wide range
of protocols based on
[ASN.1 specification](https://www.itu.int/rec/dologin_pub.asp?lang=e&id=T-REC-X.208-198811-W!!PDF-E&type=items).

**NOTE:** The package is now maintained by *Christian Heimes* and
*Simon Pichugin* in project https://github.com/pyasn1/pyasn1.

Features
--------

* Generic implementation of ASN.1 types (X.208)
* Standards compliant BER/CER/DER codecs
* Can operate on streams of serialized data
* Dumps/loads ASN.1 structures from Python types
* 100% Python, works with Python 2.7 and 3.6+
* MT-safe
* Contributed ASN.1 compiler [Asn1ate](https://github.com/kimgr/asn1ate)

Why using pyasn1
----------------

ASN.1 solves the data serialisation problem. This solution was
designed long ago by the wise Ancients. Back then, they did not
have the luxury of wasting bits. That is why ASN.1 is designed
to serialise data structures of unbounded complexity into
something compact and efficient when it comes to processing
the data.

That probably explains why many network protocols and file formats
still rely on the 30+ years old technology. Including a number of
high-profile Internet protocols and file formats.

Quite a number of books cover the topic of ASN.1. 
[Communication between heterogeneous systems](http://www.oss.com/asn1/dubuisson.html)
by Olivier Dubuisson is one of those high quality books freely 
available on the Internet.

The pyasn1 package is designed to help Python programmers tackling
network protocols and file formats at the comfort of their Python
prompt. The tool struggles to capture all aspects of a rather
complicated ASN.1 system and to represent it on the Python terms.

How to use pyasn1
-----------------

With pyasn1 you can build Python objects from ASN.1 data structures.
For example, the following ASN.1 data structure:

```bash
Record ::= SEQUENCE {
  id        INTEGER,
  room  [0] INTEGER OPTIONAL,
  house [1] INTEGER DEFAULT 0
}
```

Could be expressed in pyasn1 like this:

```python
class Record(Sequence):
    componentType = NamedTypes(
        NamedType('id', Integer()),
        OptionalNamedType(
            'room', Integer().subtype(
                implicitTag=Tag(tagClassContext, tagFormatSimple, 0)
            )
        ),
        DefaultedNamedType(
            'house', Integer(0).subtype(
                implicitTag=Tag(tagClassContext, tagFormatSimple, 1)
            )
        )
    )
```

It is in the spirit of ASN.1 to take abstract data description 
and turn it into a programming language specific form.
Once you have your ASN.1 data structure expressed in Python, you
can use it along the lines of similar Python type (e.g. ASN.1
`SET` is similar to Python `dict`, `SET OF` to `list`):

```python
>>> record = Record()
>>> record['id'] = 123
>>> record['room'] = 321
>>> str(record)
Record:
 id=123
 room=321
>>>
```

Part of the power of ASN.1 comes from its serialisation features. You
can serialise your data structure and send it over the network.

```python
>>> from pyasn1.codec.der.encoder import encode
>>> substrate = encode(record)
>>> hexdump(substrate)
00000: 30 07 02 01 7B 80 02 01 41
```

Conversely, you can turn serialised ASN.1 content, as received from
network or read from a file, into a Python object which you can
introspect, modify, encode and send back.

```python
>>> from pyasn1.codec.der.decoder import decode
>>> received_record, rest_of_substrate = decode(substrate, asn1Spec=Record())
>>>
>>> for field in received_record:
>>>    print('{} is {}'.format(field, received_record[field]))
id is 123
room is 321
house is 0
>>>
>>> record == received_record
True
>>> received_record.update(room=123)
>>> substrate = encode(received_record)
>>> hexdump(substrate)
00000: 30 06 02 01 7B 80 01 7B
```

The pyasn1 classes struggle to emulate their Python prototypes (e.g. int,
list, dict etc.). But ASN.1 types exhibit more complicated behaviour.
To make life easier for a Pythonista, they can turn their pyasn1
classes into Python built-ins:

```python
>>> from pyasn1.codec.native.encoder import encode
>>> encode(record)
{'id': 123, 'room': 321, 'house': 0}
```

Or vice-versa -- you can initialize an ASN.1 structure from a tree of
Python objects:

```python
>>> from pyasn1.codec.native.decoder import decode
>>> record = decode({'id': 123, 'room': 321, 'house': 0}, asn1Spec=Record())
>>> str(record)
Record:
 id=123
 room=321
>>>
```

With ASN.1 design, serialisation codecs are decoupled from data objects,
so you could turn every single ASN.1 object into many different 
serialised forms. As of this moment, pyasn1 supports BER, DER, CER and
Python built-ins codecs. The extremely compact PER encoding is expected
to be introduced in the upcoming pyasn1 release.

More information on pyasn1 APIs can be found in the
[documentation](http://snmplabs.com/pyasn1/),
compiled ASN.1 modules for different protocols and file formats
could be found in the pyasn1-modules 
[repo](https://github.com/pyasn1/pyasn1-modules).

How to get pyasn1
-----------------

The pyasn1 package is distributed under terms and conditions of 2-clause
BSD [license](http://snmplabs.com/pyasn1/license.html). Source code is freely
available as a GitHub [repo](https://github.com/pyasn1/pyasn1).

You could `pip install pyasn1` or download it from [PyPI](https://pypi.org/project/pyasn1).

If something does not work as expected, 
[open an issue](https://github.com/epyasn1/pyasn1/issues) at GitHub or
post your question [on Stack Overflow](https://stackoverflow.com/questions/ask)
or try browsing pyasn1 
[mailing list archives](https://sourceforge.net/p/pyasn1/mailman/pyasn1-users/).

Copyright (c) 2005-2020, [Ilya Etingof](mailto:etingof@gmail.com).
All rights reserved.


CHANGES.rst
Revision 0.5.1, released 20-11-2023
---------------------------------------

- Added support for PyPy 3.10 and Python 3.12
  [pr #32](https://github.com/pyasn1/pyasn1/pull/32/)
- Updated RTD configuration to include a dummy index.rst
  redirecting to contents.html, ensuring compatibility with
  third-party documentation and search indexes.
  [pr #47](https://github.com/pyasn1/pyasn1/pull/47/)
- Fixed the API breakage wih decoder.decode(substrateFun=...).

  A substrateFun passed to ``decoder.decode()`` can now be either
  v0.4 Non-Streaming or v0.5 Streaming. pyasn1 will detect and
  handle both cases transparently.

  A substrateFun passed to one of the new streaming decoders is
  still expected to be v0.5 Streaming only.
  [pr #30](https://github.com/pyasn1/pyasn1/pull/30/)
  [pr #39](https://github.com/pyasn1/pyasn1/pull/39/)

Revision 0.5.0, released 19-04-2023
---------------------------------------

- Change `RealEncoder.supportIndefLenMode` type to a boolean
  [pr #21](https://github.com/pyasn1/pyasn1/pull/21/)
- Fix CI for py39 test environment
  [pr #25](https://github.com/pyasn1/pyasn1/pull/25/)
- Replace all snmplabs.com links
  [issue #4](https://github.com/pyasn1/pyasn1/issues/4)
- Use correct SPDX identifier for the license
  [pr #16](https://github.com/pyasn1/pyasn1/pull/16)
- Re-add ``tagMap`` and ``typeMap`` module level attributes to all
  encoder and decoder modules. They are aliases for ``TAG_MAP`` and
  ``TYPE_MAP``, [issue #9](https://github.com/pyasn1/pyasn1/issues/9).
- Restore API for passing for ``tagMap`` and ``typeMap`` arguments
  to ``Encoder`` and ``Decoder`` classes by name and position,
  [issue #12](https://github.com/pyasn1/pyasn1/issues/12).
- Re-add ``tagMap`` and ``typeMap`` module level attributes to all
  encoder and decoder modules. They are aliases for ``TAG_MAP`` and
  ``TYPE_MAP``, [issue #9](https://github.com/pyasn1/pyasn1/issues/9).
- Restore API for passing for ``tagMap`` and ``typeMap`` arguments
  to ``Encoder`` and ``Decoder`` classes by name and position,
- Make BER/CER/DER decoders streaming and suspendible

  The goal of this change is to make the decoder yielding on input
  data starvation and resuming from where it stopped whenever the
  caller decides to try again (hopefully making sure that some more
  input becomes available).

  This change makes it possible for the decoder to operate on streams
  of data (meaning that the entire DER blob might not be immediately
  available on input).

  On top of that, the decoder yields partially reconstructed ASN.1
  object on input starvation making it possible for the caller to
  inspect what has been decoded so far and possibly consume partial
  ASN.1 data.

  All these new feature are natively available through
  `StreamingDecoder` class. Previously published API is implemented
  as a thin wrapper on top of that ensuring backward compatibility.
- Added support for Python 3.8, 3.9, 3.10, 3.11
- Removed support for EOL Pythons 2.4, 2.5, 2.6, 3.2, 3.3, 3.4, 3.5
- Added support for PyPy 3.7, 3.8, 3.9
- Modernized packaging and testing. pyasn1 now uses ``setup.cfg``,
  ``pyproject.toml``, [build](https://pypi.org/project/build/), and
  GitHub Actions.
- PyPI package ownership for `pyasn1` and `pyasn1-module` has been
  transfered to *Christian Heimes* and *Simon Pichugin* in
  [PyPI support ticket #2090](https://github.com/pypa/pypi-support/issues/2090).
- The upstream repositories for `pyasn1` and `pyasn1-modules` are now
  in the GitHub organization https://github.com/pyasn1/.

Revision 0.4.8, released 16-11-2019
-----------------------------------

- Added ability of combining `SingleValueConstraint` and
  `PermittedAlphabetConstraint` objects into one for proper modeling
  `FROM ... EXCEPT ...` ASN.1 clause.

Revision 0.4.7, released 01-09-2019
-----------------------------------

- Added `isInconsistent` property to all constructed types. This property
  conceptually replaces `verifySizeSpec` method to serve a more general
  purpose e.g. ensuring all required fields are in a good shape. By default
  this check invokes subtype constraints verification and is run by codecs
  on value de/serialisation.
- Deprecate `subtypeSpec` attributes and keyword argument. It is now
  recommended to pass `ValueSizeConstraint`, as well as all other constraints,
  to `subtypeSpec`.
- Fixed a design bug in a way of how the items assigned to constructed
  types are verified. Now if `Asn1Type`-based object is assigned, its
  compatibility is verified based on having all tags and constraint
  objects as the type in field definition. When a bare Python value is
  assigned, then field type object is cloned and initialized with the
  bare value (constraints verificaton would run at this moment).
- Added `WithComponentsConstraint` along with related
  `ComponentPresentConstraint` and `ComponentAbsentConstraint` classes
  to be used with `Sequence`/`Set` types representing
  `SET ... WITH COMPONENTS ...` like ASN.1 constructs.

Revision 0.4.6, released 31-07-2019
-----------------------------------

- Added previously missing `SET OF ANY` construct encoding/decoding support.
- Added `omitEmptyOptionals` option which is respected by `Sequence`
  and `Set` encoders. When `omitEmptyOptionals` is set to `True`, empty
  initialized optional components are not encoded. Default is `False`.
- New elements to `SequenceOf`/`SetOf` objects can now be added at any
  position - the requirement for the new elements to reside at the end
  of the existing ones (i.e. s[len(s)] = 123) is removed.
- List-like slicing support added to `SequenceOf`/`SetOf` objects.
- Removed default initializer from `SequenceOf`/`SetOf` types to ensure
  consistent behaviour with the rest of ASN.1 types. Before this change,
  `SequenceOf`/`SetOf` instances immediately become value objects behaving
  like an empty list. With this change, `SequenceOf`/`SetOf` objects
  remain schema objects unless a component is added or `.clear()` is
  called.

setup-binary.py
# This file is part of pyasn1 software.
#
# Copyright (c) 2005-2018, Ilya Etingof <etingof@gmail.com>
# License: http://snmplabs.com/pyasn1/license.html
#
# Created on 2018.06.12
#
# Author: Pedro Algarvio
#
# Copyright 2018 Giovanni Cannata
# Copyright 2018 Pedro Algarvio

# Import python libs
import os
import sys
import json
import glob
import shutil
import subprocess
import multiprocessing
import distutils.dist
# pylint: disable=no-name-in-module
from distutils.core import Command
from distutils.command.clean import clean
from distutils.command.build_py import build_py
from distutils.command.build_ext import build_ext
from distutils.command.install import install
from setuptools import setup
from setuptools.command.develop import develop


# ----- Global Variables -------------------------------------------------------------------------------------------->
# Change to source's directory prior to running any command
try:
    SETUP_DIRNAME = os.path.dirname(__file__)
except NameError:
    # We're most likely being frozen and __file__ triggered this NameError
    # Let's work around that
    SETUP_DIRNAME = os.path.dirname(sys.argv[0])
if SETUP_DIRNAME != '':
    os.chdir(SETUP_DIRNAME)

SETUP_DIRNAME = os.path.abspath(SETUP_DIRNAME)
NO_EXTENSIONS_PLACEHOLDER = object()

classifiers = """\
Development Status :: 5 - Production/Stable
Environment :: Console
Intended Audience :: Developers
Intended Audience :: Education
Intended Audience :: Information Technology
Intended Audience :: System Administrators
Intended Audience :: Telecommunications Industry
License :: OSI Approved :: BSD License
Natural Language :: English
Operating System :: OS Independent
Programming Language :: Python :: 2
Programming Language :: Python :: 2.4
Programming Language :: Python :: 2.5
Programming Language :: Python :: 2.6
Programming Language :: Python :: 2.7
Programming Language :: Python :: 3
Programming Language :: Python :: 3.2
Programming Language :: Python :: 3.3
Programming Language :: Python :: 3.4
Programming Language :: Python :: 3.5
Programming Language :: Python :: 3.6
Topic :: Communications
Topic :: Software Development :: Libraries :: Python Modules
"""

params = {
    'name': 'pyasn1-binary',
    'version': open(os.path.join(SETUP_DIRNAME, 'pyasn1', '__init__.py')).read().split('\'')[1],
    'description': 'ASN.1 types and codecs',
    'long_description': 'Pure-Python implementation of ASN.1 types and DER/BER/CER codecs (X.208)',
    'maintainer': 'Ilya Etingof <etingof@gmail.com>',
    'author': 'Ilya Etingof',
    'author_email': 'etingof@gmail.com',
    'url': 'https://github.com/etingof/pyasn1',
    'platforms': ['any'],
    'classifiers': [x for x in classifiers.split('\n') if x],
    'license': 'BSD',
    'packages': ['pyasn1',
                 'pyasn1.type',
                 'pyasn1.compat',
                 'pyasn1.codec',
                 'pyasn1.codec.ber',
                 'pyasn1.codec.cer',
                 'pyasn1.codec.der',
                 'pyasn1.codec.native']}
# Change to source's directory prior to running any command
try:
    SETUP_DIRNAME = os.path.dirname(__file__)
except NameError:
    # We're most likely being frozen and __file__ triggered this NameError
    # Let's work around that
    SETUP_DIRNAME = os.path.dirname(sys.argv[0])
if SETUP_DIRNAME != '':
    os.chdir(SETUP_DIRNAME)

SETUP_DIRNAME = os.path.abspath(SETUP_DIRNAME)
# <---- Global Variables ---------------------------------------------------------------------------------------------


# ----- Custom Distutils/Setuptools Commands ------------------------------------------------------------------------>
class BuildPy(build_py):

    def find_package_modules(self, package, package_dir):
        modules = build_py.find_package_modules(self, package, package_dir)
        for package, module, filename in modules:
            if self.distribution.develop_mode is False:
                if package == 'pyasn1.codec.ber' and module in ('encoder',):
                    # We have issues when cython compiling asn1, include it as a python module
                    yield package, module, filename
                    continue
            if self.distribution.develop_mode is False and module not in ('__init__',):
                # We only want __init__ and _version python files
                # All others will be built as extensions
                continue
            yield package, module, filename


class BuildExt(build_ext):

    def run(self):
        self.extensions = self.distribution.get_ext_modules()
        build_ext.run(self)

    def build_extensions(self):
        if int(self.distribution.concurrency) > 1:
            self.check_extensions_list(self.extensions)
            import multiprocessing.pool
            multiprocessing.pool.ThreadPool(
                processes=int(self.distribution.concurrency)).map(
                    self.build_extension, self.extensions)
        else:
            build_ext.build_extensions(self)


class Clean(clean):

    def finalize_options(self):
        self.distribution._ext_modules = None
        clean.finalize_options(self)

    def run(self):
        clean.run(self)
        for subdir in ('pyasn1', 'tests'):
            root = os.path.join(os.path.dirname(__file__), subdir)
            for dirname, dirs, _ in os.walk(root):
                for to_remove_filename in glob.glob('{0}/*.py[ocx]'.format(dirname)):
                    os.remove(to_remove_filename)
                for to_remove_filename in glob.glob('{0}/*.c'.format(dirname)):
                    os.remove(to_remove_filename)
                for to_remove_filename in glob.glob('{0}/*.so'.format(dirname)):
                    os.remove(to_remove_filename)
                for dir_ in dirs:
                    if dir_ == '__pycache__':
                        shutil.rmtree(os.path.join(dirname, dir_))


class Develop(develop):

    def finalize_options(self):
        self.distribution.develop_mode = True
        develop.finalize_options(self)

    def run(self):
        # Clean up any previously generated cython files
        self.run_command('clean')
        develop.run(self)


class Install(install):

    def finalize_options(self):
        install.finalize_options(self)
        if self.distribution.has_ext_modules():
            self.install_lib = self.install_platlib

    def run(self):
        self.run_command('build_ext')
        install.run(self)


class BuildWheels(Command):

    description = 'Build manylinux1 wheel packages'
    user_options = []

    def initialize_options(self):
        '''
        Abstract method that is required to be overwritten
        '''

    def finalize_options(self):
        '''
        Abstract method that is required to be overwritten
        '''

    def run(self):
        try:
            docker_binary = subprocess.check_output('which docker', shell=True).decode(sys.stdin.encoding or 'asccii').strip()
        except subprocess.CalledProcessError:
            print('Docker binary not found')
            sys.exit(1)

        for manylinux_docker_image in ('quay.io/pypa/manylinux1_x86_64',
                                       'quay.io/pypa/manylinux1_i686'):
            if manylinux_docker_image.endswith('i686'):
                pre_cmd = 'linux32'
            else:
                pre_cmd = ''
            dest = manylinux_docker_image.replace('quay.io/pypa/manylinux1_', '')
            try:
                subprocess.check_call('{} pull {}'.format(docker_binary, manylinux_docker_image), shell=True)
            except subprocess.CalledProcessError as exc:
                print('Failed to download docker image {}: {}'.format(manylinux_docker_image, exc))
                sys.exit(1)
            try:
                subprocess.check_call(
                    '{docker_binary} run --rm -v `pwd`:/io {docker_image} {pre_cmd} '
                    '/io/build-binary-wheels.sh {dest} `id -u` `id -g`'.format(
                        docker_binary=docker_binary,
                        docker_image=manylinux_docker_image,
                        pre_cmd=pre_cmd,
                        dest=dest
                    ),
                    shell=True
                )
            except subprocess.CalledProcessError as exc:
                print('Failed to build package with docker image {}: {}'.format(manylinux_docker_image, exc))
                sys.exit(1)

        try:
            packages_listing = subprocess.check_output('ls wheelhouse/', shell=True).decode(sys.stdin.encoding or 'asccii')
            print('Built Packages')
            print(packages_listing)
        except subprocess.CalledProcessError as exc:
            print('Failed to list built packages: {}'.format(exc))
            sys.exit(1)
# <---- Custom Distutils/Setuptools Commands -------------------------------------------------------------------------


# ----- Custom Distribution Class ----------------------------------------------------------------------------------->
class Distribution(distutils.dist.Distribution):
    '''
    Distribution class
    '''

    DEFAULT_EXT_MODULES = object()
    global_options = distutils.dist.Distribution.global_options + [
        ('concurrency=', None, 'Parallelize extension module builds')
    ]

    def __init__(self, attrs=None):
        distutils.dist.Distribution.__init__(self, attrs)
        self._ext_modules = NO_EXTENSIONS_PLACEHOLDER
        self.concurrency = multiprocessing.cpu_count()
        self.develop_mode = False
        self.name = params['name']
        self.version = params['version']
        self.description = params['description']
        self.long_description = params['long_description']
        self.author = params['author']
        self.author_email = params['author_email']
        self.url = params['url']
        self.cmdclass.update(
            {
                'clean': Clean,
                'build_py': BuildPy,
                'build_ext': BuildExt,
                'install': Install,
                'build_wheels': BuildWheels,
                'develop': Develop,
            }
        )
        self.license = params['license']
        self.packages, self.package_dir, self.package_data = self.discover_packages()
        self.zip_safe = False
        self.update_metadata()

    def update_metadata(self):
        for attrname in dir(self):
            if attrname.startswith('__'):
                continue
            if attrname == 'ext_modules':
                # Don't trigger cythonize
                continue
            attrvalue = getattr(self, attrname, None)
            if attrvalue == 0:
                continue
            if hasattr(self.metadata, 'set_{0}'.format(attrname)):
                getattr(self.metadata, 'set_{0}'.format(attrname))(attrvalue)
            elif hasattr(self.metadata, attrname):
                try:
                    setattr(self.metadata, attrname, attrvalue)
                except AttributeError:
                    pass

    def find_ext(self):
        from Cython.Distutils.extension import Extension
        ret = []
        for package in ('pyasn1',):
            for root, _, files in os.walk(os.path.join(SETUP_DIRNAME, package)):
                commonprefix = os.path.commonprefix([SETUP_DIRNAME, root])
                for filename in files:
                    full = os.path.join(root, filename)
                    if not filename.endswith(('.py', '.c')):
                        continue
                    if filename in ('__init__.py',):
                        continue
                    relpath = os.path.join(root, filename).split(commonprefix)[-1][1:]
                    if relpath in ('pyasn1/codec/ber/encoder.py',):
                        # We have issues when cython compiling asn1, skipt it
                        continue
                    module = os.path.splitext(relpath)[0].replace(os.sep, '.')
                    ret.append(Extension(module, [full]))
        return ret

    def discover_packages(self):
        modules = []
        pkg_data = {}
        pkg_dir = {}
        for package in ('pyasn1',):
            for root, _, files in os.walk(os.path.join(SETUP_DIRNAME, package)):
                if '__init__.py' not in files:
                    continue
                pdir = os.path.relpath(root, SETUP_DIRNAME)
                modname = pdir.replace(os.sep, '.')
                modules.append(modname)
                pkg_data.setdefault(modname, []).append('*.so')
                pkg_dir[modname] = pdir
        return modules, pkg_dir, pkg_data

    def get_ext_modules(self):
        if self.develop_mode is False and self._ext_modules is NO_EXTENSIONS_PLACEHOLDER:
            from Cython.Build import cythonize
            self.ext_modules = self._ext_modules = cythonize(self.find_ext(),
                                                             nthreads=int(self.concurrency))
        return self._ext_modules

    # ----- Static Data -------------------------------------------------------------------------------------------->
    @property
    def _property_classifiers(self):
        return params['classifiers']

    #@property
    #def _property_dependency_links(self):
    #    return []

    #@property
    #def _property_tests_require(self):
    #    return ['pytest']
    # <---- Static Data ----------------------------------------------------------------------------------------------

    # ----- Dynamic Data -------------------------------------------------------------------------------------------->
    @property
    def _property_package_data(self):
        return self.package_data

    @property
    def _property_data_files(self):
        return []

    @property
    def _property_install_requires(self):
        return [i.strip() for i in open('requirements.txt').readlines()]

    @property
    def _property_extras_require(self):
        return {}

    @property
    def _property_scripts(self):
        return []

    @property
    def _property_setup_requires(self):
        return []

    #@property
    #def _property_entry_points(self):
    #    return {'console_scripts': ['...']}
    # <---- Dynamic Data ---------------------------------------------------------------------------------------------

    # ----- Overridden Methods -------------------------------------------------------------------------------------->
    def parse_command_line(self):
        args = distutils.dist.Distribution.parse_command_line(self)
        # Setup our property functions after class initialization and
        # after parsing the command line since most are set to None
        # ATTENTION: This should be the last step before returning the args or
        # some of the requirements won't be correctly set
        for funcname in dir(self):
            if not funcname.startswith('_property_'):
                continue
            property_name = funcname.split('_property_', 1)[-1]
            setattr(self, property_name, getattr(self, funcname))

        return args

    def has_ext_modules(self):
        # Make sure we tell distutils that this is not a pure python package
        return self.develop_mode is False

    def has_pure_modules(self):
        # Make sure we tell distutils that this is a pure python package
        return True
    # <---- Overridden Methods ---------------------------------------------------------------------------------------

# <---- Custom Distribution Class ------------------------------------------------------------------------------------
if __name__ == '__main__':
    setup(distclass=Distribution)
